src/pathfinder/yapf/yapf_base.hpp

```diff
@@ -249,10 +249,6 @@ class CYapfBaseT {
 			return;
 		}
 
-		if (m_max_search_nodes > 0 && (m_pBestIntermediateNode == nullptr || (m_pBestIntermediateNode->GetCostEstimate() - m_pBestIntermediateNode->GetCost()) > (n.GetCostEstimate() - n.GetCost()))) {
-			m_pBestIntermediateNode = &n;
-		}
-
```

Comment:
Shouldn't this code, though with `*openNode` instead of `n`, also be used in the branch below where it calls `InsertOpenNode` (line 266/262)? It feels like `*openNode` could then be the best intermediate node in that branch. Granted, doing it on `n` is definitely wrong here and there.
Or in other words, shouldn't this "is this the best intermediate node logic" be called just after both calls to `InsertOpenNode`?

Reply:
Yeah, this is one possibility. My main issue with this was that if you reach a new best intermediate node and find out it's already in the open or closed list, you can only "fix" this in the case of an existing worse node in the open list. So I opted for setting the best intermediate node only when the node is truly fresh, in order to avoid surprises.

But if you think it's worthwhile to also adjust in the case of a worse open node, I can adjust.

Reply:
I suppose this really just depends on whether the best intermediate node is supposed to consider any node or only nodes that end up in the open/closed list.

Reply:
I think the only way you can get to this branch of the code is with smaller initial segments that have uneven amount of signals which are then followed by a long, uninterrupted segments that end due to ESRB_SEGMENT_TOO_LONG. I've managed to trigger a breakpoint in this part of code when I was testing something unrelated on the savegame from #9271 (happens to trains around 931x702), but it took me quite a while to actually construct a situation where this changes the tracks the vehicles take.

In pretty much all non-artificial situations I've tried, the pathfinder simply extends the best ESRB_SEGMENT_TOO_LONG node, sets a new best intermediate node and the problem disappears (since ESRB_SEGMENT_TOO_LONG nodes generally pass enough signals to get past the lookahead).

---

I suppose saying that the behavior would be surprising might be overstating it somewhat. Basically, when I opened the PR I was thinking about `m_pBestIntermediateNode` in the same way as `m_pBestDestNode`, i.e. it should be any node closest to the destination, even if it never ends up in the A* graph.

The best destination node can also end up in a similar situation (i.e. never being added to the open list), but it notifies the node list not to reuse it. While this could also be used to fix this problem, it seems extremely wasteful as most intermediate nodes are then thrown away. It would also result in some other annoyances with, for example, branch pruning.

So, we're left with the possibility of setting the best intermediate node in 2 out of the 4 cases that can occur (fresh node, worse open, better open, closed) and anything we do here is going to be slightly wrong and thus, in some sense, surprising.

On the other hand, if `m_pBestIntermediateNode` is only supposed to contain nodes that actually end up in the A* graph, then only handling the fresh node and worse open node would be correct. After thinking about it some more, I'm actually liking this second option a bit better, but I also haven't been around when YAPF was added so perhaps that interpretation is wrong.

Reply:
I'm now convinced that doing the latter (i.e. only consider nodes that end up in the A* graph) is the better option.